draft-ietf-mimi-content.md

```diff
@@ -1184,6 +1212,26 @@ Clients compliant with this specification must be able to download
 ExternalParts with `http` and `https` URLs, and decrypt downloaded content
 encrypted with the AES-128-GCM AEAD algorithm.
 
+### Specifics of Github Flavored Markdown in MIMI
+
+A MIMI content client supports GitHub Flavored Markdown as defined in [@!GFM],
+with one exception (no raw HTML support) and a fixed list of supported
+extensions, as further described in the bullet points below.
+This document specifies what can be sent inside a MIMI content message; it does
+not restrict or prescribe in any way how input from a user is interpreted by an
+Instant Messaging client that support MIMI, before any message resulting from
+that input is sent.
+
+Note that rendering Markdown as plain text is an aceptable form of "support".
+
+- Raw HTML MUST NOT be included other than blank HTML comments `<!-- -->`.
```

Comment:
Why do we want to support hiding content using HTML comments in messaging?

This will be rendered entirely different in clients that just display the markdown as plain text (which is allowed as written above), so using comments could be used as a weird way to display different content to different clients.

I'm of course aware that the multipart/alternative semantics also allow for this, but that is intentional and clearly specified, this one seems less intentional and is easier to miss.

Some context: In XMPP there is language based alternative selection of the message body. It's a rarely used festure, but some clients will display an indication and even allow selection of the language when they receive a multi-language message. This makes it possible to spot messages that have different content in their different alternatives, which is important for example in the context of moderation.

Reply:
If you look at the GFM spec, it is the only way to separate two otherwise valid types of markup. If you search in the GFM spec for the comment string with a single space, you will find it.
It is unfortunate, if the comment is displayed raw, but it would not be sent frequently, and it would be an annoyance, but should not interfere with the human understanding the rest of the content.

Reply:
Is this a common variant of GFM? Are there parsers/processors available that support this variant? Or are we just about to create a new variant of Markdown incompatible with everything that's already out there?


The two examples in the GFM spec could both also be realized without the `<!-- -->`.

<details><summary>Example 288</summary>

```markdown
- foo
- bar

* baz
* bim
```

- foo
- bar

* baz
* bim

</details> 

<details><summary>Example 289</summary>

~~~markdown
-   foo

    notcode

-   foo
```
code
```
~~~

-   foo

    notcode

-   foo
```
code
```

</details> 

I would even go as far as saying that the variant without empty HTML comment is more readable  when printed in plain than the variant with empty HTML comment.

I'm not saying there aren't going to be things that can't be done in GFM without using those empty HTML comments - but I'm also certain that there are things that can't be done in GFM even with those, so we seem to have settled on accepting these weird restrictions that come from using a Markdown language.

Reply:
I had mixed results with your suggestions..

(This is **not** a **variant** of GFM, it is **the spec** for GFM, which of course is implemented by GitHub.)

For Example 289, using triple-tilde worked to interrupt a list with a code block:

1) a
2) b

```
code
```

1) 1 (new list)
2) 2

But for Example 288, using different punctuation characters for an unordered list did not create a new list for me, and that does not cover the case of multiple ordered lists (I used an HTML comment here as a separator, but `<div/>` would havealso  worked too):

1) a
2) b

<!-- -->

1) x
2) y

Reply:
With "variant" I was referring to the variant of GFM you're proposing in this document: GFM, with the three GFM extensions listed, but disallowing raw HTML except the empty HTML comment by displaying the HTML as verbatim code (by replacing `<` with `&lt;`). GFM clearly allows raw HTML and there is an extension to filter some tags to make it safe, but filtering all tags and non-empty HTML comments is behavior noncompliant with the GFM spec. As such using the `text/markdown;variant=GFM` content type for this is inappropriate. I also question if there's any markdown implementation out there that allows to be configured exactly as per this specification, and if the answer is no, I would say that specifying our own variant here is probably a bad idea, because it will inevitably result in people using markdown processors that divert from the specification.

Using different punctuation for list items clearly creates a new list in GFM, but I'm pretty sure there are other markdown variants where it doesn't. See [section 5.4](https://github.github.com/gfm/#lists) of GFM spec:
> A [list](https://github.github.com/gfm/#list) is a sequence of one or more list items [of the same type](https://github.github.com/gfm/#of-the-same-type). The list items may be separated by any number of blank lines.
>
> Two list items are [of the same type](https://github.github.com/gfm/#of-the-same-type) if they begin with a [list marker](https://github.github.com/gfm/#list-marker) of the same type. Two list markers are of the same type if (a) they are bullet list markers using the same character (-, +, or *) or (b) they are ordered list numbers with the same delimiter (either . or )).

It also explains that, as there are two valid delimiters for ordered lists, it does work for them as well.

<details><summary>Example 288 for ordered lists</summary>
<p>

```
1. a
2. b

1) x
2) y
```

1. a
2. b

1) x
2) y

</p>
</details>

Reply:
How would you feel about defining `text/markdown;variant=GFM-noHTML` (no HTML comment or other PIs either)?

I could register this variant in the variant registry.

Reply:
At MIMI interim agreed we will go forward with a new variant GFM-MIMI that has no HTML.